### PR TITLE
add DynamoDBTarget

### PR DESCRIPTION
**Pull request by ouanixi**
## Description
Added a DynamoDB Target object. This is a simple target in the style of the postgres Target. It creates a marker-table if it doesn't exist and pushes items with date and update_id

Users can add a [dynamodb] setup entry in the config file defining the name of the marker-table and the throughput config required for the creation of the marker table.

Default values:
marker-table: table_updates
read-capacity-units: 1
write-capacity-units: 1

## Have you tested this? If so, how?
Unit tests included. And used it in production

### CODE REVIEW

**Comment by ouanixi** [comment]
Wow what just happened with the CI :) I've run the tests locally as recommended:
    tox -e flake8
    tox -e py27-nonhdfs
and everything runs OK.
I've added boto3 to the tox.ini file but I have a feeling this is not the way I should do it (not too familiar with tox and travis so I'd appreciate some pointers)

**Comment by dlstadther** [comment]
@ouanixi Do you know what's happening here with these tests?

**Comment by ouanixi** [comment]
@dlstadther think this one is ready to be reviewed :)

File: .travis.yml [hunk 1]
```diff
@@ -15,6 +15,9 @@ env:
     - BQ_TEST_PROJECT_ID=luigi-travistestenvironment
     - BQ_TEST_INPUT_BUCKET=luigi-bigquery-test
     - GOOGLE_APPLICATION_CREDENTIALS=test/gcloud-credentials.json
+    - AWS_DEFAULT_REGION=us-east-1
```
Comment (dlstadther) [inline .travis.yml]: us-east-1 is defined here, but eu-west-1 is used within dynamodb_test.py...

File: test/contrib/dynamodb_test.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+from helpers import unittest
+from luigi.contrib.dynamodb import DynamoDBTarget
+from moto import mock_dynamodb2
+
+
+@mock_dynamodb2
+class TestDynamoClient(unittest.TestCase):
+
+    client = boto3.resource('dynamodb', region_name='eu-west-1')
```
Comment (dlstadther) [inline test/contrib/dynamodb_test.py]: Differs from Travis default... why?

Comment (ouanixi) [reply test/contrib/dynamodb_test.py]: No particular reason, the client is mocked anyway. I'll remove the region from the client construction and let it resolve itself from the ENV var (as per .travis.yml)

File: .travis.yml (hunk 1 above)
Comment (ouanixi) [reply .travis.yml]: Yep that's just a lack of consistency from my part. I'll remove the region from dynamodb_test.py and let it resove itself from the env vars.

File: luigi/contrib/dynamodb.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    def exists(self):
+        table = self.resource.Table(self.marker_table)
+
+        try:
+            response = table.get_item(Key={'update_id': self.update_id})
+        except ClientError as e:
+            if e.response['Error']['Code'] == 'ResourceNotFoundException':
+                return False
+            else:
+                raise
```
[the remark this answers is not in the document]
Comment (ouanixi) [reply luigi/contrib/dynamodb.py]: The idea here is that I want to propagate any exception that is not 'ResourceNotFoundException' to the client code.  We just want to know if the item exists so I think any other exception should be handled by the user.

List of possible exceptions can be found here:
http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html#Programming.Errors.MessagesAndCodes

File: luigi/contrib/dynamodb.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    write_capacity_units = luigi.configuration.get_config().get(
+        'dynamodb', 'write-capacity-units', 1)
+
+    def __init__(self, resource, update_id):
+        """
+        Args:
+            resource (DynamoDB.ServiceResource): created using boto3.resource()
+            update_id (str): An identifier for this dataset
+        """
+        self.update_id = update_id
```
[the remark this answers is not in the document]
Comment (ouanixi) [reply luigi/contrib/dynamodb.py]: Not sure what you mean Dillon. As far as I can see, both RedshiftTarget and PostgresTarget use the same pattern (i.e update_id is required)

https://github.com/spotify/luigi/blob/master/luigi/contrib/postgres.py#L134

[the remark this answers is not in the document]
Comment (dlstadther) [reply luigi/contrib/dynamodb.py]: Ah. Looks like i mixed things up a bit. I was thinking about how the update_id is optional for the Copy classes of Redshift/Postgres. They both inherit their [update_id from rdbms](https://github.com/spotify/luigi/blob/master/luigi/contrib/rdbms.py#L99); which defaults to their task_id

[the remark this answers is not in the document]
Comment (dlstadther) [reply luigi/contrib/dynamodb.py]: This can be ignored then

**Comment by dlstadther** [comment]
@ouanixi Do you have any colleagues or someone else with domain knowledge to review this?

**Comment by ouanixi** [comment]
@dlstadther hmm it might be a bit tricky. I don't know many people that use Luigi. I'll see what I can do.

**Pull request closed without merging**